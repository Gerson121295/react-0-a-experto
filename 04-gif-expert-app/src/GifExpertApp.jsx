import { useState } from "react"
import AddCategory from "./components/AddCategory";
import GifGrid from "./components/GifGrid";

const GifExpertApp = () => {

    //hook useState, categories es el estado y con setCategories se cambia el estado de categories, usestate(['One Punch]) inicializa categories con 'One Punch'
    const [categories, setCategories] = useState(['One Punch']);


    //FUncion agregar nueva categoria
    const onAddCategory = (newCategory) => {

        if(categories.includes(newCategory)) return; //Valida si el array categories ya incluye el nueva categoria newCategory si ya lo tiene hace un return;

       //console.log('Valoran')
       setCategories([newCategory, ...categories]); //Forma1: se agrega la nueva categoria al inicio newCategory (recibida del hijo AddCategory) y con ...categories  se hace un clon de las categorias que se tenian anteriores, 
       //setCategories([...categories, 'Valoran']); //Forma1: con: ...categories  se hace un clon de las categorias que se tenian anteriores, luego se agrega la nueva 'Valoran'
        //setCategories(cat => [...cat, 'Valorant']); //forma 2: una funcion define callback cat, la cual se le hace un clon de las categorias anteriores luego se agrega la nueva 'Valoran'
    }


  return (
    <>
      <h1>GifExpertAp</h1>

    {/* Input */}
    <AddCategory 
        // Forma1: Al component Hijo AddCategory se llama para rederizarse se le envia la prop setCategories{setCategories} la cual trae la categoria que el usuario escribio en input para buscar el gif
        //setCategories={setCategories}

        //Forma 2: Al component Hijo AddCategory se llama para rederizarse se le envia la prop onNewCategory la cual trae la categoria que el usuario escribio en input para buscar el gif
        //onNewCategory={(value)=> onAddCategory(value)} //cuando a una funcion recibe un argumento y este argumento se envia a la funcion se omite y se escribe solo el nombre de la funcion:
        onNewCategory={onAddCategory}
    />


    {/* Listado de Gif */}
        {
            categories.map( (category) => (  //map recorre el arreglo para imprimir cada item
                <GifGrid 
                    key={category}
                    category={category}
                />
            ))
        }


    {/* Gif Item */}

    </>
  )
}

export default GifExpertApp
