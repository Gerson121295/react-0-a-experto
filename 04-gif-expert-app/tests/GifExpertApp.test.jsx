import { render, screen, fireEvent, waitFor } from "@testing-library/react"
import GifExpertApp from "../src/GifExpertApp"



describe('Prueba al component ', () => {

    const newCategory = 'Dragon Ball'
 
    test('debe hacer match con el snapshot', () => {

        const { container} = render( <GifExpertApp /> );
        expect( container ).toMatchSnapshot();
 
    });

    /*
    test('debe agregar una categoria', () => {     
        // Renderizar el componente
        render( <GifExpertApp /> );
    
        // Obtener el input y el form
        const input = screen.getByRole('textbox');
        const form = screen.getByRole('form');
    
        // Agregar la primera categoría
        fireEvent.input(input, { target: { value: 'New Category' } });
        fireEvent.submit(form);
    
        // Limpiar el valor del input
        fireEvent.change(input, { target: { value: '' } });
    
        // Agregar la segunda categoría
        fireEvent.input(input, { target: { value: 'Another Category' } });
        fireEvent.submit(form);
    
        // Verificar que solo se haya agregado una nueva categoría
        const categoryElements = screen.getAllByTestId('category');
        expect(categoryElements.length).toBe(1);
    });
*/

test('no debe agregar una nueva categoria si la categoria ya existe', () => {
    const { container } = render(<GifExpertApp />);

    const input = screen.getByRole('textbox');
    const form = screen.getByRole('form');

    fireEvent.input(input, { target: { value: newCategory }});
    fireEvent.submit(form);

    waitFor(() => {
        const cardGrids = container.querySelectorAll('.card-grid');
        expect(cardGrids.length).toBe(1);
    });
});

});