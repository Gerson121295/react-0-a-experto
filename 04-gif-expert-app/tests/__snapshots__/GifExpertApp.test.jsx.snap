// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba al component  debe hacer match con el snapshot 1`] = `
<div>
  <h1>
    GifExpertAp
  </h1>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar gifs"
      type="text"
      value="One Punch"
    />
  </form>
  <h3>
    One Punch
  </h3>
  <h2>
    Cargando...
  </h2>
  <div
    class="card-grid"
  />
</div>
`;
